libraries/Microsoft.Bot.Builder.Azure/AzureTableStorage.cs

```diff
@@ -3,164 +3,369 @@
+                        var propValue = prop.GetValue(entity);
+                        if (propValue == null) return;
 
-            public StoreItemEntity() { }
+                        if (IsSimple(propValue.GetType()))
+                        {
+                            properties.Add(prefix + prop.Name, EntityProperty.CreateEntityPropertyFromObject(propValue));
+                        }
+                        else
+                        {
+                            var typeQualifiedName = propValue.GetType().AssemblyQualifiedName;
+                            properties.Add(prop.Name + "__type", EntityProperty.GeneratePropertyForString(typeQualifiedName));
+                            Flatten(properties, propValue, prop.Name + "_");
+                        }
+                    }
 
-            public StoreItemEntity(EntityKey key, object obj)
-                : this(key.PartitionKey, key.RowKey, obj)
-            { }
+                    return;
+                }
+                else
+                {
+                    // POCO objects
+                    var childProperties = EntityPropertyConverter.Flatten(entity, new OperationContext());
+                    foreach (var childProp in childProperties)
+                    {
+                        properties.Add(prefix + childProp.Key, childProp.Value);
+                    }
+                }
+            }
 
-            public StoreItemEntity(string partitionKey, string rowKey, object obj)
-                : base(partitionKey, rowKey)
+            private static object CreateInstaceOf(Type type)
             {
-                this.ETag = (obj as IStoreItem)?.eTag;
-                this.Json = JsonConvert.SerializeObject(obj, Formatting.None, serializationSettings);
+                var ctor = type
+                    .GetConstructors()
+                    .FirstOrDefault(c => c.GetParameters().Length > 0);
+
+                return ctor != null
+                    // Possible anonymous type
+                    ? ctor.Invoke
+                        (ctor.GetParameters()
+                            .Select(p =>
+                                p.HasDefaultValue ? p.DefaultValue :
+                                p.ParameterType.IsValueType && Nullable.GetUnderlyingType(p.ParameterType) == null
+                                    ? Activator.CreateInstance(p.ParameterType)
+                                    : null
+                            ).ToArray()
+                        )
+                    : Activator.CreateInstance(type);
             }
 
-            public string Json { get; set; }
+            private static bool IsSimple(Type type)
+            {
+                return type.IsPrimitive
+                  || type.IsEnum
+                  || type.Equals(typeof(string))
+                  || type.Equals(typeof(decimal));
+            }
 
-            public object AsObject()
+            private static bool IsAnonymousType(Type type)
```

Comment:
> NOTE: all comments on anonymous type support apply... I don't think we should do it.

Pretty sure you grabbed this off StackOverflow, but I would make some changes.

First I would make it an expression bodied method simply because it is a good candidate for it.

Next, we should never use `.Count()` when you don't need to know exactly how many there are. It's a _potential_ `O(N)` cost. Instead you should prefer `.Any()` when you're just trying to figure out `0` or not. Now, it's true the implementation of `Count()` will try to optimize for `ICollection<T>`  types to use their `.Count` property, which in this case, because `GetCustomAttributes` returns an array, will short cut since array implements `ICollection<T>`. However, that raises the next question: why we're are we even using `.Count()` at all when we are working with an array has a `.Length` property of its own that we can just check directly?

Next, the call to `String::Contains` to check for `"AnonymousType"` will be cheaper than calling `GetCustomAttributes` every time, so I'd rearrange the two.

Implementation proposed based on these comments:

```CSharp
private static bool IsAnonymousType(Type type) => 
    type.FullName.Contains("AnonymousType") && type.GetCustomAttributes(typeof(CompilerGeneratedAttribute), false).Length;
```
Finally, I would argue that doing this check _every single time an instance of state is deserialized_ is costly and that we should memoize the results into a `ConcurrentDictionary<Type, bool>` for subsequent checks to be a simple `O(1)` lookup. 

> NOTE: Again, I would only do this if we're truly going to keep support for anonymous types and it could certainly be done as a follow up optimization PR.